library/src/main/java/org/xmtp/android/library/frames/FramesClient.kt

```diff
@@ -0,0 +1,92 @@
+package org.xmtp.android.library.frames
+
+import android.util.Base64
+import org.xmtp.android.library.Client
+import org.xmtp.android.library.frames.FramesConstants.PROTOCOL_VERSION
+import org.xmtp.android.library.messages.PrivateKeyBuilder
+import org.xmtp.android.library.messages.Signature
+import org.xmtp.android.library.messages.getPublicKeyBundle
+import org.xmtp.proto.message.contents.PublicKeyOuterClass.SignedPublicKeyBundle
+import java.security.MessageDigest
+import org.xmtp.proto.message.contents.Frames.FrameActionBody
+import org.xmtp.proto.message.contents.Frames.FrameAction
+
+class FramesClient(private val xmtpClient: Client, var proxy: OpenFramesProxy = OpenFramesProxy()) {
+
+    suspend fun signFrameAction(inputs: FrameActionInputs): FramePostPayload {
+        val opaqueConversationIdentifier = buildOpaqueIdentifier(inputs)
+        val frameUrl = inputs.frameUrl
+        val buttonIndex = inputs.buttonIndex
+        val inputText = inputs.inputText
+        val state = inputs.state
+        val now = System.currentTimeMillis() / 1000
+        val frameActionBuilder = FrameActionBody
+            .newBuilder()
+            .setFrameUrl(frameUrl)
+            .setButtonIndex(buttonIndex)
+            .setOpaqueConversationIdentifier(opaqueConversationIdentifier)
+            .setTimestamp(now)
+            .setUnixTimestamp(now.toInt())
+
+        if (inputText != null) {
+            frameActionBuilder.inputText = inputText
+        }
+        if (state != null) {
+            frameActionBuilder.state = state
+        }
+
+        val toSign = frameActionBuilder.build()
+        val signedAction = Base64.encodeToString(buildSignedFrameAction(toSign), Base64.NO_WRAP)
+
+        val untrustedData = FramePostUntrustedData(frameUrl, now, buttonIndex, inputText, state, xmtpClient.address, opaqueConversationIdentifier, now.toInt())
+        val trustedData = FramePostTrustedData(signedAction)
+
+        return FramePostPayload("xmtp@$PROTOCOL_VERSION", untrustedData, trustedData)
+    }
+
+    private suspend fun signDigest(digest: ByteArray): Signature {
+        val signedPrivateKey = xmtpClient.keys.identityKey
+        val privateKey = PrivateKeyBuilder.buildFromSignedPrivateKey(signedPrivateKey)
+        return PrivateKeyBuilder(privateKey).sign(digest)
+    }
+
+    private fun getPublicKeyBundle(): SignedPublicKeyBundle {
+        return xmtpClient.keys.getPublicKeyBundle()
+    }
+
+    private suspend fun buildSignedFrameAction(actionBodyInputs: FrameActionBody): ByteArray {
+        val digest = sha256(actionBodyInputs.toByteArray())
+        val signature = signDigest(digest)
+
+        val publicKeyBundle = getPublicKeyBundle()
+        val frameAction = FrameAction.newBuilder()
+            .setActionBody(actionBodyInputs.toByteString())
+            .setSignature(signature)
+            .setSignedPublicKeyBundle(publicKeyBundle).build()
```

Comment:
Same here. Setters are a very Java way of doing things. Kotlin prefers this method

```suggestion
        val frameAction = FrameAction.newBuilder().also {
            it.actionBody = actionBodyInputs.toByteString()
            it.signature = signature
            it.signedPublicKeyBundle = publicKeyBundle
         }.build()
```